container/monitor.go

```diff
@@ -1,46 +1,123 @@
 package container
 
 import (
+	"sync"
 	"time"
 
 	"github.com/Sirupsen/logrus"
+	containertypes "github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/container/stream"
+	"github.com/docker/docker/daemon/logger"
+	"github.com/docker/docker/restartmanager"
+	"golang.org/x/net/context"
 )
 
 const (
 	loggerCloseTimeout = 10 * time.Second
 )
 
-// Reset puts a container into a state where it can be restarted again.
-func (container *Container) Reset(lock bool) {
-	if lock {
-		container.Lock()
-		defer container.Unlock()
+var containerState = &runStateStore{s: make(map[string]*runState)}
+
+type runStateStore struct {
+	mu sync.Mutex
+	s  map[string]*runState
+}
+
+func getRunState(c *Container) *runState {
+	containerState.mu.Lock()
+	s := containerState.s[c.ID]
+	containerState.mu.Unlock()
+	return s
+}
+
+func removeRunState(c *Container) {
+	containerState.mu.Lock()
+	delete(containerState.s, c.ID)
+	containerState.mu.Unlock()
+}
+
+func addRunState(c *Container, s *runState) {
```

Comment:
looks like `add` is not really used as `add`, more like `create`. Maybe remove `s` argument and allocate state here?